torch/distributed/_tensor/dispatch.py

```diff
@@ -88,6 +91,14 @@ def dispatch(
         """
         Main dispatching logic
         """
+        # Redistribute grad_output tensor to the same placement as input tensor
+        args = list(args)
```

Comment:
hmmm I think we shouldn't do those checks and run redistribute, this is really hacky. I would prefer if this thing can be implemented as  sharding propagation + resharding workflow. If you want to those custom hacks instead of supporting sharding propagation directly. we should not hijack to op dispatching logic like this.

So if you want to keep the current customization, you can probably do it in this way:
1. register `aten.convolution_backward.default` and `aten.convolution.default` to the custom_op_handler in the `OpDispatcher` https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/dispatch.py#L78
2. the custom handler for these two conv ops should call into the `tp_conv.py`, where inside tp_conv.py, you can put these custom logic into the custom dispatch functions(i.e. tp_convolution, tp_convolution_backward), where you could do all custom things like unwrapping DTensor to get local shards, run conv rule propagation, run tp_conv, re-wrapping local outputs with DTensors.

Reply:
@wanchaol It is the problem of `sharding propagation + resharding` of `torch.Tensor.mean`, which undermines the `Shard(3)` placement of activations of convolution.

![image](https://github.com/pytorch/pytorch/assets/70677813/aa7ee208-21dd-41ee-9526-0f0e95eb0027)

If the sharding propagation of mean([-2,-1]) can be fixed properly, I think this hijack can be eliminated. The convolution per se does not require any custom hacks actually,

Reply:
Copy my comment over:

Thanks for the super detailed explaination! The shardings for the fwd and bwd for a specific operator does not have to be aligned as it is mostly determined by the backward formula as you observed, that's why we have the resharding logics to make sure if the backward ops are sharding propagated differently it would perform resharding when necessary, currently using OpStrategy we can specify a op only "accepts" Shard(3) so that if the argument is not Shard(3) it would perform argument resharding to Shard(3) first. That's why I was suggesting we should try to make the ops uses the Op Strategy approach. Additionally, we essentially dispatch convolution to tp_convolution and tp_convolution_bwd call instead of using the original aten.convolution for local shard computation, this is not a uniform way as other ops.

But if you want to land the currently way of doing sharding propagation for convolution ops without the resharding logic, I am fine too, but we need to avoid polluting the dispatch.py, I think you can register the custom_handler to custom handle both convolution and convolution backward

Reply:
The tp_conv custom handler we can register should be relatively simple, after register a custom handler, you just need to do the following things in the handler:
1. unwrap arguments from DTensor to local_shard tensor and the sharding specs, this code can mostly copy from the dispatch.py
2. run conv and conv_backward sharding propagation rule to get the output sharding
3. run tp_conv and tp_conv_backward
4. rewrap the outputs to DTensor with the output sharding, this part can just call the OpDispatcher.wrap

Reply:
I see, and the first two steps seem to introduce duplicated code, is it good? Also, which place is better to put `aten_convolution_handler `and `aten_convolution_backward_handler `, `dispatch.py` or `tp_conv.py`?

Reply:
If switching to `OpStrategy` for convolution can make cleaner code, I prefer to go for this. Do you have some suggestions on required changes?

Reply:
I decided to keep using `register_prop_rule` and use custom handlers in this PR. Following the above 4 steps, I have this code snippet:

```
def convolution_backward_handler(
    op_call: torch._ops.OpOverload,
    args: Tuple[object, ...],
    kwargs: Dict[str, object],
) -> object:
    # Redistribute grad_output tensor to the same placement as input tensor
    args = list(args)
    if op_call == torch.ops.aten.convolution_backward.default:
        assert isinstance(args[0], dtensor.DTensor) and isinstance(
            args[1], dtensor.DTensor
        )
        args[0] = args[0].redistribute(args[1].device_mesh, args[1].placements)
    args = tuple(args)
    local_tensor_args = [
        arg.to_local() if isinstance(arg, dtensor.DTensor) else arg for arg in args
    ]
    local_tensor_args = cast(Tuple[object, ...], local_tensor_args)
    local_kwargs = kwargs
    local_results = tp_convolution_backward(op_call, local_tensor_args, local_kwargs)
  
    # HOW CAN I GET output_sharding WITHOUT DUPLICATING TOO MUCH CODE???

    return OpDispatcher.wrap(local_results, output_sharding.output_spec)
```

The remaining question is how to get  `output_sharding ` in this handler function without duplicating too much code. Any suggestions are appreciated.

Reply:
I have added customer handlers as suggested. Thank you so much for refactoring work to make components of op dispatch more usable.